src/WebJobs.Extensions.DurableTask/EntityScheduler/MessageSorter.cs

```diff
@@ -0,0 +1,243 @@
+
+        [JsonProperty(DefaultValueHandling = DefaultValueHandling.Ignore)]
+        public Dictionary<string, DateTime> Sent { get; set; }
+
+        [JsonProperty(DefaultValueHandling = DefaultValueHandling.Ignore)]
+        public Dictionary<string, ReceiveBuffer> Received { get; set; }
+
+        [JsonProperty(DefaultValueHandling = DefaultValueHandling.Ignore)]
+        public DateTime ReceiveHorizon { get; set; }
+
+        [JsonProperty(DefaultValueHandling = DefaultValueHandling.Ignore)]
+        public DateTime SendHorizon { get; set; }
+
+        public int NumberBufferedRequests =>
+            this.Received?.Select(kvp => kvp.Value.Buffered?.Count ?? 0).Sum() ?? 0;
+
+        /// <summary>
+        /// Called on the sending side, to fill in timestamp and predecessor fields.
+        /// </summary>
+        public void LabelOutgoingMessage(RequestMessage message, string destination, DateTime now, TimeSpan reorderWindow)
+        {
+            DateTime timestamp = now;
+
+            if (this.SendHorizon + reorderWindow + MinIntervalBetweenCollections < now)
+            {
+                this.SendHorizon = now - reorderWindow;
+
+                // clean out send clocks that are past the reorder window
+
+                if (this.Sent != null)
+                {
+                    List<string> expired = null;
+
+                    foreach (var kvp in this.Sent)
+                    {
+                        if (kvp.Value < this.SendHorizon)
+                        {
+                            (expired ?? (expired = new List<string>())).Add(kvp.Key);
+                        }
+                    }
+
+                    if (expired != null)
+                    {
+                        foreach (var t in expired)
+                        {
+                            this.Sent.Remove(t);
+                        }
+
+                        expired.Clear();
+                    }
+
+                    if (this.Sent.Count == 0)
+                    {
+                        this.Sent = null;
+                    }
+                }
+            }
+
+            if (this.Sent == null)
+            {
+                this.Sent = new Dictionary<string, DateTime>();
+            }
+            else if (this.Sent.TryGetValue(destination, out var last))
+            {
+                message.Predecessor = last;
+
+                // ensure timestamps are monotonic even if system clock is not
+                if (timestamp <= last)
+                {
+                    timestamp = new DateTime(last.Ticks + 1);
+                }
+            }
+
+            message.Timestamp = timestamp;
+            this.Sent[destination] = timestamp;
+        }
+
+        /// <summary>
+        /// Called on the receiving side, to reorder and deduplicate within the window.
+        /// </summary>
+        public IEnumerable<RequestMessage> ReceiveInOrder(RequestMessage message, TimeSpan reorderWindow)
```

Comment:
I am not sure I understand the question.

Reply:
I notice we are using the `yield` pattern here instead of returning an `List` or some other concrete implementation. This normally indicates that you want to lazily retrieve messages only when we need them. However, it looks like the only places that consume this method handles the request messages in batches anyways, so I'm not sure what advantages `yield` offers in this case.

Reply:
The point was to avoid the indirection of first creating a `List` container, filling it, and then enumerating it again just once.

Using `yield` allows me to process the enumerated messages directly without creating an intermediate single-use list object.